### ISSUE TITLE
fail to cross-compile linux->windows: clang-cl rejects `-xc`

### ISSUE THREAD

**Issue by tanriol** [nabijaczleweli/rust-embed-resource#68]
I've been using rust-embed-resource via `tauri` for quite some time by cross-compiling the Windows version on Linux using `clang-cl` as compiler (cargo `config.toml` snippet below).

```
# Defined based on https://jake-shadle.github.io/xwin/
[target.x86_64-pc-windows-msvc]
linker = "lld-link"
rustflags = [
    "-Lnative=/path/to/xwin/crt/lib/x86_64",
    "-Lnative=/path/to/xwin/sdk/lib/um/x86_64",
    "-Lnative=/path/to/xwin/sdk/lib/ucrt/x86_64",
]

[env]
CC_x86_64_pc_windows_msvc = "clang-cl"
CXX_x86_64_pc_windows_msvc = "clang-cl"
AR_x86_64_pc_windows_msvc = "llvm-lib"
```

For some time the build used to succeed with a warning about `-xc` being unrecognized. Now (presumably after clang/llvm upgrade, although I'm not sure) the build fails with the following error: `clang-cl: error: unsupported option '-x c'; did you mean '/TC' or '/TP'?`

<details>
<summary>More context</summary>

```
The following warnings were emitted during compilation:                                                              
                                                                                                                     
warning: test-cross@0.0.0: clang-cl: error: unsupported option '-x c'; did you mean '/TC' or '/TP'?                                                                                                                                        
                                                                                                                     
error: failed to run custom build command for `test-cross v0.0.0 (/tmp/test-cross/src-tauri)`                                                                                                                                              
                                                                                                                     
Caused by:                                                                                                                                                                                                                                 
  process didn't exit successfully: `/tmp/test-cross/src-tauri/target/debug/build/test-cross-35b858db60db2b87/build-script-build` (exit status: 1)                                                                                         
  --- stdout                                                                                                         
  cargo:rerun-if-env-changed=TAURI_CONFIG                                                                                                                                                                                                  
  cargo:rerun-if-changed=tauri.conf.json                                                                             
  cargo:rustc-cfg=desktop                                                                                                                                                                                                                  
  cargo:rustc-env=TAURI_ANDROID_PACKAGE_PREFIX=com_tauri_dev                                         
  cargo:rustc-cfg=dev                                                                                                                                                                                                                      
  cargo:PERMISSION_FILES_PATH=/tmp/test-cross/src-tauri/target/x86_64-pc-windows-msvc/debug/build/test-cross-18f49f7db6edd6fc/out/app-manifest/__app__-permission-files                                                                    
  cargo:rerun-if-changed=capabilities                                                                                                                                                                                                      
  cargo:rustc-env=TAURI_ENV_TARGET_TRIPLE=x86_64-pc-windows-msvc                                       
  package.metadata does not exist                                                                                    
  cargo:warning=clang-cl: error: unsupported option '-x c'; did you mean '/TC' or '/TP'?               
                                                                                                                     
  --- stderr                                                                                                         
                                                                                                                     
                                                                                                                     
  error occurred: Command "clang-cl" "-nologo" "-MD" "-Z7" "-Brepro" "-m64" "-I" "/tmp/test-cross/src-tauri/target/x86_64-pc-windows-msvc/debug/build/test-cross-18f49f7db6edd6fc/out" "-W4" "-xc" "-DRC_INVOKED" "-E" "--" "/tmp/test-cros
s/src-tauri/target/x86_64-pc-windows-msvc/debug/build/test-cross-18f49f7db6edd6fc/out/resource.rc" with args clang-cl did not execute successfully (status code exit status: 1).                                                           
                                                                                                                     
                                                                                                                     
warning: build failed, waiting for other jobs to finish...    
```

</details>

I've traced this to [this `cc` invocation](https://github.com/nabijaczleweli/rust-embed-resource/blob/master/src/non_windows.rs#L91-L96).

**Comment by nabijaczleweli** [comment]
indeed
```rust
            CompilerType::LlvmRc { has_no_preprocess } => {
                let preprocessed_path = format!("{}/{}-preprocessed.rc", out_dir, prefix);
                fs::write(&preprocessed_path,
                          apply_macros_cc(cc::Build::new().define("RC_INVOKED", None), macros)
                              .flag("-xc")
                              .file(resource)
                              .cargo_metadata(false)
                              .include(out_dir)
                              .expand())
                    .unwrap();
```

**Comment by nabijaczleweli** [comment]
Indeed, and ever-so-helpfully I don't see a generalised way to pass `-x`-equivalents in https://docs.rs/cc/latest/cc/struct.Build.html :v 

Can you try making the line with `.flag("-xc")` be `.flag("-Xclang").flag("-xc")` instead? That would turn this into `Tool::is_like_msvc()`.

**Comment by wfeii1980** [comment]
clang 18, this issue directly causes an error rather than a warning

```shell
clang-cl --version
clang version 18.1.8
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: /usr/bin
```

```shell
cargo:warning=clang-cl: error: unsupported option '-x c'; did you mean '/TC' or '/TP'?

  --- stderr


  error occurred: Command "clang-cl" "-nologo" "-MT" "-O1" "-Brepro" "-m64" "-I" "/home/wii/data/works2/test_sdk/target/x86_64-pc-windows-msvc/release/build/test-29949c93bf72f0f6/out" "--target=x86_64-pc-windows-msvc" "-Wno-unused-command-line-argument" "-fuse-ld=lld-link" "/imsvc/opt/.xwin_cache/xwin/crt/include" "/imsvc/opt/.xwin_cache/xwin/sdk/include/ucrt" "/imsvc/opt/.xwin_cache/xwin/sdk/include/um" "/imsvc/opt/.xwin_cache/xwin/sdk/include/shared" "-xc" "-DRC_INVOKED" "-E" "--" "assets/test-manifest.rc" with args clang-cl did not execute successfully (status code exit status: 1).
```

**Comment by nabijaczleweli** [comment]
Try applying the same diff as above and see if it fixes it :)

**Comment by wfeii1980** [comment]
Why do we need the '-xc' parameter

default clang-cl = 18
```shell
> clang-cl -xc -DRC_INVOKED -E assets/demo-manifest.rc 
clang-cl: error: unsupported option '-x c'; did you mean '/TC' or '/TP'?

> clang-cl -Xclang -xc -DRC_INVOKED -E assets/demo-manifest.rc 
# 1 "assets/demo-manifest.rc"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 419 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "assets/demo-manifest.rc" 2

1 24 "demo.exe.manifest"

MAINICON ICON "app-icon.ico"

> clang-cl -DRC_INVOKED -E assets/demo-manifest.rc 
# 1 "assets/demo-manifest.rc"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 419 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "assets/demo-manifest.rc" 2

1 24 "demo.exe.manifest"

MAINICON ICON "app-icon.ico"

> /opt/clang-17/bin/clang-cl -xc -DRC_INVOKED -E assets/demo-manifest.rc 
clang-cl: warning: unknown argument ignored in clang-cl: '-xc' [-Wunknown-argument]
# 1 "assets/demo-manifest.rc"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 412 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "assets/demo-manifest.rc" 2

1 24 "demo.exe.manifest"

MAINICON ICON "app-icon.ico"

> /opt/clang-17/bin/clang-cl -Xclang -xc -DRC_INVOKED -E assets/demo-manifest.rc 
# 1 "assets/demo-manifest.rc"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 412 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "assets/demo-manifest.rc" 2

1 24 "demo.exe.manifest"

MAINICON ICON "app-icon.ico"

> /opt/clang-17/bin/clang-cl -DRC_INVOKED -E assets/demo-manifest.rc 
# 1 "assets/demo-manifest.rc"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 412 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "assets/demo-manifest.rc" 2

1 24 "demo.exe.manifest"

MAINICON ICON "app-icon.ico"
```

**Comment by wfeii1980** [comment]
I saw it https://github.com/nabijaczleweli/rust-embed-resource/issues/47#issue-1292141546

`-xc` It is required for cc (gcc)

WSL env It is `compiler.is_like_gnu() = true` ?

**Comment by nabijaczleweli** [comment]
Can you try the current `master` branch (at least e98e9aa28d66b0c2ff43516b41847cac372310dd)?

**Comment by tanriol** [comment]
Sorry for a late response. The master branch works for me.

**Comment by nabijaczleweli** [comment]
Released in [`v2.4.3`](https://github.com/nabijaczleweli/rust-embed-resource/releases/tag/v2.4.3), backported to [`v1.6.13`](https://github.com/nabijaczleweli/rust-embed-resource/releases/tag/v1.6.13).
